### PR TITLE
[FAB-16937] complete by hooking CC fwork with ext. CC

### PR DESCRIPTION
**Pull request by muralisrini**
With the external builder and chaincode support prepared for
chaincode-as-server model this task finally fills in the
remaining pieces to complete the model
. adjust chaincode support for client gRPC stream
. fork runtime to connect to cc server if server info
 is provided

Change-Id: I09ecb17fcc762e2cb5e706176bec9b6fa738962e
Signed-off-by: muralisr <srinivasan.muralidharan99@gmail.com>

### CODE REVIEW

**Review by guoger — COMMENTED** [review]
overall LGTM, just some nits

File: core/chaincode/extcc/extcc_handler_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+				}
+				err := i.Run("ccid", ccinfo, shandler)
+				Expect(err).To(BeNil())
+			})
+		})
+		When("chaincode info is nil", func() {
+			It("returns an error", func() {
+				ccinfo = nil
+				err := i.Run("ccid", ccinfo, shandler)
+				//Expect(err).To(MatchError(ContainSubstring("cannot start connector without connection properties")))
```
[the remark this answers is not in the document]
Comment (guoger) [reply core/chaincode/extcc/extcc_handler_test.go]: i mean line 89 only... not sure why it shows 4 lines...

File: core/chaincode/extcc/extcc_handler.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+// createConnection - standard grpc client creating using ClientConfig info (surprised there isn't
+// a helper method for this)
+func (i *ExternalChaincodeRuntime) createConnection(ccid string, ccinfo *ccintf.ChaincodeServerInfo) (*grpc.ClientConn, error) {
+	if ccinfo == nil {
+		return nil, errors.New("cannot start connector without connection properties")
+	}
+
+	grpcClient, err := comm.NewGRPCClient(ccinfo.ClientConfig)
+	if err != nil {
+		return nil, errors.WithMessage(err, fmt.Sprintf("error creating grpc client to %s", ccid))
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler.go]: will do

File: core/chaincode/extcc/extcc_handler.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	conn, err := grpcClient.NewConnection(ccinfo.Address)
+	if err != nil {
+		return nil, errors.WithMessage(err, fmt.Sprintf("error creating grpc connection to %s", ccinfo.Address))
+	}
+
+	extccLogger.Debugf("created external chaincode connection: %s", ccid)
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler.go]: will do

File: core/chaincode/extcc/extcc_handler_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler_test.go]: will do

File: core/chaincode/extcc/extcc_handler_test.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+					Certificate:       []byte("fake-cert"),
+					Key:               []byte("fake-key"),
+					ServerRootCAs:     [][]byte{[]byte("fake-root-cert")},
+				},
+				Timeout: 10 * time.Second,
+			},
+		}
+	})
+
+	Context("Start", func() {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler_test.go]: Actually it should be `Context("Run", `. I'll fix that... but I think that was not your point. I think there's some value in calling out we are testing "Run" (even if it is the only exported function)? Also, would be easier to organize if we add more in future.

File: core/chaincode/extcc/extcc_handler_test.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+			ClientConfig: comm.ClientConfig{
+				SecOpts: comm.SecureOptions{
+					UseTLS:            true,
+					RequireClientCert: true,
+					Certificate:       []byte("fake-cert"),
+					Key:               []byte("fake-key"),
+					ServerRootCAs:     [][]byte{[]byte("fake-root-cert")},
+				},
+				Timeout: 10 * time.Second,
+			},
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler_test.go]: True, will move it to error case.

File: core/chaincode/extcc/extcc_handler_test.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+					ClientConfig: comm.ClientConfig{
+						KaOpts:  comm.DefaultKeepaliveOptions,
+						Timeout: 10 * time.Second,
+					},
+				}
+				err := i.Run("ccid", ccinfo, shandler)
+				Expect(err).To(BeNil())
+			})
+		})
+		When("chaincode info is nil", func() {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler_test.go]: will do

**Review by jyellick — COMMENTED** [review]
I think this is getting pretty close, but a number of comments.

File: core/chaincode/extcc/extcc_handler.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	HandleChaincodeStream(stream ccintf.ChaincodeStream) error
+}
+
+type ExternalChaincodeRuntime struct {
+}
+
+// createConnection - standard grpc client creating using ClientConfig info (surprised there isn't
+// a helper method for this)
+func (i *ExternalChaincodeRuntime) createConnection(ccid string, ccinfo *ccintf.ChaincodeServerInfo) (*grpc.ClientConn, error) {
+	if ccinfo == nil {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler.go]: It should not be nil but as its passed from an external package,  just a bit of defensive programming.

File: core/chaincode/extcc/extcc_handler.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func (i *ExternalChaincodeRuntime) Run(ccid string, ccinfo *ccintf.ChaincodeServerInfo, sHandler StreamHandler) error {
+	extccLogger.Debugf("Starting external chaincode connection: %s", ccid)
+	conn, err := i.createConnection(ccid, ccinfo)
+	if err != nil {
+		return errors.WithMessagef(err, "error cannot create connection for %s", ccid)
+	}
+
+	//remove connection after processing (or error)
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler.go]: Will remove.

File: core/chaincode/extcc/extcc_handler_test.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+			AfterEach(func() {
+				if ccserv != nil {
+					ccserv.Stop()
+				}
+				if cclist != nil {
+					cclist.Close()
+				}
+			})
+
+			It("runs to completion", func() {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler_test.go:59]: Assertions involving the mock stream are good to add. Will do.

File: core/chaincode/extcc/extcc_handler_test.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+							Key:               []byte("fake-key"),
+							ServerRootCAs:     [][]byte{[]byte("fake-root-cert")},
+						},
+						Timeout: 10 * time.Second,
+					},
+				}
+			})
+			When("chaincode info is nil", func() {
+				It("returns an error", func() {
+					ccinfo = nil
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler_test.go]: Will move initializations to BeforeEach.

File: core/chaincode/lifecycle/cache_test.go [hunk 11]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/hyperledger/fabric/common/chaincode"
 	commonledger "github.com/hyperledger/fabric/common/ledger"
 	"github.com/hyperledger/fabric/common/util"
+	extccmock "github.com/hyperledger/fabric/core/chaincode/extcc/mock"
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/lifecycle/cache_test.go]: Will remove the mock and use nil.

File: core/chaincode/runtime_launcher.go [hunk 12]
```diff
@@ -71,16 +79,26 @@ func (r *RuntimeLauncher) Launch(ccid string) error {
...
-			var ccservinfo *ccintf.ChaincodeServerInfo
+			// go through the build process to obtain connecion information
 			ccservinfo, err := r.Runtime.Build(ccid)
 			if err != nil {
 				startFailCh <- errors.WithMessage(err, "error building chaincode")
 				return
 			}
+
+			// chaincode server model indicated... proceed to connect to CC
 			if ccservinfo != nil {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/runtime_launcher.go:89]: Just that the caller and callee are in separate packages and the method could get called in future in ways not anticipated.

File: core/chaincode/lifecycle/custodian.go [hunk 13]
```diff
@@ -70,7 +71,7 @@ func (cc *ChaincodeCustodian) Close() {
 	cc.cond.Signal()
 }
 
-func (cc *ChaincodeCustodian) Work(buildRegistry *container.BuildRegistry, builder ChaincodeBuilder, launcher ChaincodeLauncher) {
+func (cc *ChaincodeCustodian) Work(buildRegistry *container.BuildRegistry, builder ChaincodeBuilder, launcher ChaincodeLauncher, shandler extcc.StreamHandler) {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/lifecycle/custodian.go]: Good idea. Will add a shim to peer (only user of this).

File: core/chaincode/extcc/extcc_handler.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+	if err != nil {
+		return nil, errors.WithMessagef(err, "error creating grpc connection to %s", ccinfo.Address)
+	}
+
+	extccLogger.Debugf("Created external chaincode connection: %s", ccid)
+
+	return conn, nil
+}
+
+func (i *ExternalChaincodeRuntime) Run(ccid string, ccinfo *ccintf.ChaincodeServerInfo, sHandler StreamHandler) error {
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler.go]: More of a Stream really.... more like the standard bidi "Chat". Will use `Stream`.

File: core/chaincode/runtime_launcher.go [hunk 15]
```diff
@@ -71,16 +79,26 @@ func (r *RuntimeLauncher) Launch(ccid string) error {
...
-				startFailCh <- errors.New("peer as client to be implemented")
+				if err = r.ConnectionHandler.Run(ccid, ccservinfo, streamHandler); err != nil {
+					startFailCh <- errors.WithMessage(err, fmt.Sprintf("connection to %s failed", ccid))
```
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/runtime_launcher.go]: Will fix.

File: core/chaincode/extcc/extcc_handler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jyellick) [reply core/chaincode/extcc/extcc_handler.go]: If a function takes a pointer as an argument, it's expected not to be nil (unless the function makes it very clear that 'nil' is an acceptable input.  There are lots of examples in the go standard library of this:

https://golang.org/src/net/http/fs.go#L672
https://golang.org/src/archive/tar/writer.go#L67
https://golang.org/src/encoding/json/encode.go#L195

And many more.  When I see a nil check like this, the obvious implication is that "Sometimes, the caller should call this function with 'nil'", but that's not the case.

**Review by jyellick — DISMISSED** [review]

File: core/chaincode/extcc/extcc_handler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (muralisrini) [reply core/chaincode/extcc/extcc_handler.go]: Point taken, this is an "internal" api after all, passing nil is a bug and better crash. Will fix.

**Review by jyellick — DISMISSED** [review]

File: core/chaincode/chaincode_support.go [hunk 16]
```diff
@@ -42,7 +43,7 @@ type Runtime interface {
 
 // Launcher is used to launch chaincode runtimes.
 type Launcher interface {
-	Launch(ccid string) error
+	Launch(ccid string, streamHandler extcc.StreamHandler) error
```
Comment (jyellick) [inline core/chaincode/chaincode_support.go:46]: We could simplify this by using the custodian shim here too.

**Review by jyellick — APPROVED** [review]

File: core/chaincode/chaincode_support.go (hunk 16 above)
Comment (muralisrini) [reply core/chaincode/chaincode_support.go:46]: Assuming I didn't misunderstand, the custodian shim is an adapter for the Launcher interface.. we cannot change the adapted ?

Comment (jyellick) [reply core/chaincode/chaincode_support.go:46]: Ah, nevermind this comment, my mistake.

**Review by denyeart — APPROVED** [review]

**Pull request merged**